src/lib/y2packager/release_notes_reader.rb

```diff
@@ -11,170 +11,133 @@
   class ReleaseNotesReader
     include Yast::Logger
 
+    # Product to get release notes for
+    attr_reader :product
+
     # Constructor
     #
+    # @param product             [Product]           Product to get release notes for
     # @param release_notes_store [ReleaseNotesStore] Release notes store to cache data
-    def initialize(release_notes_store = nil)
+    def initialize(product, release_notes_store = nil)
       @release_notes_store = release_notes_store
+      @product = product
     end
 
+    # Fallback language
+    FALLBACK_LANG = "en".freeze
+
     # Get release notes for a given product
     #
-    # Release notes are downloaded and extracted to work directory.  When
-    # release notes for a language "xx_XX" are not found, it will fallback to
-    # "xx".
-    #
-    # @param product   [Y2Packager::Product] Product
-    # @param user_lang [String]              Release notes language (falling back to "en_US"
-    #                                        and "en")
-    # @param format    [Symbol]              Release notes format (:txt or :rtf)
-    # @return [String,nil] Release notes or nil if a release notes were not found
+    # @param user_lang [String] Release notes language (falling back to "en")
+    # @param format    [Symbol] Release notes format (:txt or :rtf)
+    # @return [String,nil] Release notes or nil if release notes were not found
     #   (no package providing release notes or notes not found in the package)
-    def release_notes_for(product, user_lang: "en_US", format: :txt)
-      package = release_notes_package_for(product)
-      if package.nil?
-        log.info "No package containing release notes for #{product.name} was found"
-        return nil
+    def release_notes(user_lang: "en_US", format: :txt)
+      readers =
+        # registered system: get relnotes from RPMs and fallback to relnotes_url property
+        if registered?
+          [
+            ReleaseNotesRpmReader.new(product),
+            ReleaseNotesUrlReader.new(product)
+          ]
+        else # unregistered system: try relnotes first and fallback to RPMs
+          [
+            ReleaseNotesUrlReader.new(product),
+            ReleaseNotesRpmReader.new(product)
+          ]
+        end
+
+      readers.each do |reader|
+        rn = release_notes_via_reader(reader, user_lang, format, FALLBACK_LANG)
+        return rn if rn
       end
 
-      from_store = release_notes_store.retrieve(product.name, user_lang, format, package.version)
+      nil
+    end
+
+    # Get release notes for a given product using a reader instance
+    #
+    # @param reader        [ReleaseNotesRpmReader,ReleaseNotesUrlReader] Release notes reader
+    # @param user_lang     [String] User preferred language (falling back to fallback_lang)
+    # @param format        [Symbol] Release notes format (:txt or :rtf)
+    # @param fallback_lang [String] Release notes fallback language
```

Comment:
I see this triple of user_lang, format, fallback_lang in so many places that I think it might be better to encapsulate it in a ContentPreference object?

Reply:
Yes, it makes sense.